### ISSUE TITLE
`is` swallows type errors in `try`

### ISSUE THREAD

**Issue by arnetheduck** [nim-lang/Nim#22008]
### Description

```nim
template detect(v: untyped) =
  when typeof(v) is void:
    raiseAssert "no it's not"

detect:
  try:
    raise (ref ValueError)()
  except ValueError:
    42
```

As a variation on #22007, it turns out the compile error is silently dropped in a `when` and the code compiles just fine albeit with unexpected runtime behavior.

### Nim Version

1.6.12

### Current Output

```text
Error: unhandled exception: no it's not [AssertionDefect]
```
```


### Expected Output

_No response_

### Possible Solution

_No response_

### Additional Information

_No response_

**Comment by Araq** [comment]
But the type of your construct is `void`. It works exactly as it should.

**Comment by arnetheduck** [comment]
> It works exactly as it should.

not really - it hides a type mismatch between the branches of try/except which should be a compile-time error _before_ `is` is allowed to be evaluated.

**Comment by Araq** [comment]
Ok, but that makes it a duplicate of https://github.com/nim-lang/Nim/issues/22007

**Comment by arnetheduck** [comment]
> duplicate

I thought about that, but it means that a part of the type checking is simply not performed at all for the `except` clauses in the `when` evaluation which feels like a distinct bug (what if it was other types involved?) - I'm not familiar enough with the inner workings to tell if it's a duplicate or not. basically, #22007 gives a compile-time error and since this code has the same structure, arguably it should also fail type-checking.

**Comment by SirOlaf** [comment]
Also seems to be fixed by #22314

**Comment by ringabout** [comment]
A test case would be appreciated.
